### PR TITLE
Pluginbase as object (version 2)

### PR DESCRIPTION
**Pull request by flynd**
plugin.

The exported structure contain a pointer to a factory function the PluginManager can call to create a plugin instance. The plugin instance must inherit from either Licq::GeneralPlugin or Licq::ProtocolPlugin which specify the interface for communicating between the plugin and Licq as well as other plugins. This inheritance allows future extensions to the plugin interface to be made without having to export additional symbols.

This should also allow plugins to communicate directly with each other. For example, if a protocol plugin has a public header file, a GUI plugin could cast the plugin pointer to the protocol plugins type and access additional features not supported by the Licq daemon. (This might be a way to simplify the separation of the ICQ protocol from the Licq daemon without adding generic support for all ICQ specific features.)

The exported structure also contains the Licq version the plugin was built for. Plugin loading is only allowed if the Licq version matches with the daemon. This should catch differences between the plugin and the api that could lead to strange behavior or crashes.
 It is assumed that the last digit in the version is only used for minor releases that doesn't change the api and plugin loading will be allowed between for example 1.6.0 and 1.6.2, but not between 1.6.0 and 1.7.0.

Changes from previous branch:
- Parameters for Plugin constructor are wrapped to not expose the parameters and internal datatypes in the public headers.
- Plugin data includes a reaper function that makes plugin objects be deleted from the plugin library.

### CODE REVIEW

File: licq/include/licq/plugin.h [hunk 1]
```diff
@@ -20,123 +20,141 @@
...
+ * functions and the protected functions called from the protocol manager
+ * can be made from any thread. It is the responsibility of the plugin to make
+ * sure these functions are thread safe when needed.
  */
 class Plugin : private boost::noncopyable
 {
 public:
+  class Params;
+
+  typedef boost::error_info<struct tag_errinfo_symbol_name, std::string> errinfo_symbol_name;
```
Comment (erijo) [inline licq/include/licq/plugin.h]: This is no longer needed. It was used to add the name of the symbol that failed to load to the exception thrown from Plugin::loadSymbol().

File: licq/include/licq/pluginbase.h [hunk 2]
```diff
@@ -20,90 +20,94 @@
...
+ * This struct contains the initial data and functions needed by Licq
+ * to load a plugin.
+ */
+struct GeneralPluginData
+{
+  /**
+   * Magic value to identify an Licq plugin
+   * Must contain the characters 'L', 'i', 'c', 'q'
+   */
+  char licqMagic[4];
```
Comment (erijo) [inline licq/include/licq/pluginbase.h:42]: Should the magic be different for the two plugin types?

Comment (flynd) [inline licq/include/licq/pluginbase.h:42]: Maybe it should... The exported symbols are different, I thought that would be sufficient. But if there is a risk of abuse or mistakes unless the magic value is different, then we'll make them different.

File: licq/include/licq/plugin.h (hunk 1 above)
Comment (flynd) [inline licq/include/licq/plugin.h]: Alright, I'll drop it.

**Comment by erijo** [comment]
Besides maybe doing something about the friend declarations for the test classes I say merge it.

**Pull request closed without merging**